### ISSUE TITLE
[Bug] Fully Bayesian Inference with Pyro Does Not Work

### ISSUE THREAD

**Issue by jacobrgardner** [cornellius-gp/gpytorch#1578]
# 🐛 Bug

Right now, the way our Bayesian inference with pyro works is that we loop over all `Parameter`s with priors, call `pyro.sample` and load the sample into the parameter value. Here's an extremely simplified model that is equivalent to what we do that does NOT work:
```python
def pyro_model2(x, y):
    raw_lengthscale = pyro.sample('raw_lengthscale', Normal(0, 1))
    raw_outputscale = pyro.sample('raw_outputscale', Normal(0, 1))

    # This is not an okay way to load the results of pyro.sample statements into a Module
    covar_module.initialize(**{'raw_outputscale': raw_outputscale, 'base_kernel.raw_lengthscale': raw_lengthscale})
    
    covar_x = covar_module(x).evaluate()
    pyro.sample("obs", pyro.distributions.MultivariateNormal(torch.zeros(y.size(-1)), covar_x + torch.eye(y.size(-1))), obs=y)
```
See the attached python notebook that demonstrates that this leads to completely wrong gradients for the potential energy, so HMC is entirely broken.

The best way I've found to fix this is to use the (possibly deprecated?) `pyro.random_module` primitive. Here's a pyro model using a full GPyTorch GP that gets correct potential derivatives:
```python
def pyro_model5(x, y):
    priors= {
        'covar_module.base_kernel.raw_lengthscale': Normal(0, 1).expand([1, 1]),
        'covar_module.raw_outputscale': Normal(0, 1),
        'likelihood.noise_covar.raw_noise': Normal(0, 1).expand([1]),
        'mean_module.constant': Normal(0, 1),
   }
    fn = pyro.random_module("model", model, prior=priors)
    sampled_model = fn()
    
    output = sampled_model.likelihood(sampled_model(x))
    pyro.sample("obs", output, obs=y)
```
Now, we could definitely just replace our existing GPyTorch interface to basically do the above instead for any parameter that has a prior registered. The only problem is that I don't know if this will let us place priors over functions of `Parameters` (i.e., if we want to place a prior over the `lengthscale` rather than the `raw_lengthscale`).

Thoughts? I feel like the above is *almost* there, but it'd be great to still support placing priors over derived values of the parameters.

cc/ @rmgarnett

also cc/ @fritzo @eb8680 @martinjankowiak in case any of the pyro devs hopefully have a thought about how to do this properly that I'm just missing?

[Sampling Bug.ipynb.txt](https://github.com/cornellius-gp/gpytorch/files/6328496/Sampling.Bug.ipynb.txt)

**Comment by sdaulton** [comment]
Cc @jpchen @Balandat

**Comment by neerajprad** [comment]
I was looking into this. I think the issue here is that HMC/NUTS in Pyro expects a functional, i.e. one needs to be able to take gradients with respect to the arguments of the potential function. In this case, or more generally, when `nn.Module`s are involved and we replace the `nn.Parameter` values in the modules by replacing the `data` attribute, the graph gets disconnected, because `nn.Parameter`s are supposed to be leaf variables in the autograd graph. This should throw a `RuntimeError` which I suspect is being swallowed somewhere, likely somewhere here: https://github.com/cornellius-gp/gpytorch/blob/481f2856092681908d046acc1263aa0bc5f86877/gpytorch/module.py#L106.

This is the reason why snippet 1 does not work. As to why this works for the deprecated `pyro.random_module` is that it replaces these `nn.Parameter`s with actual tensors. While this might work for this use case, it has other unexpected side-effects, and is is one of the reasons why we do not encourage its broad use. If this works for your use for the time being, you can also replicate the same effect by using the following method in gpytorch's `Module` class to reset the parameters of the kernel (instead of calling `initialize`). This is exactly what `random_module` is doing under the hood (hacky!):

```python
  def reset_params_pyro(self, **params):
      for name, value in params.items():
          # find the innermost module
          names = name.rsplit(".")
          if len(names) > 1:
              mod_name, param_name = names
              mod = operator.attrgetter(mod_name)(self)
          else:
              mod, param_name = self, name
          # delete paramater (otherwise we get a TypeError when replacing Parameter -> Tensor)
          old_param = getattr(mod, param_name)
          delattr(mod, param_name)
          # XXX: replaces nn.Param by torch.tensor
          setattr(mod, param_name, value.expand(old_param.shape))
```

This mismatch is not an uncommon case and I wonder if something like https://github.com/facebookresearch/higher could be used to functionalize the module objects in a way that works with HMC. But the easier (and certain) route would be to build better support for `nn.Module` classes within Pyro's HMC, e.g. if we know that the PE function is a `nn.Module` we can dispatch it to a separate `potential_grad` function that is specialized for `nn.Module`.

**Comment by jacobrgardner** [comment]
@neerajprad Thanks! This is actually super helpful. At a minimum, the `reset_params_pyro` solution is much better for us than directly using `pyro.random_module` because it's obvious how it allows us to continue placing priors over transformed versions of the Parameters, and is only equally hacky.

Based on what you've said, I'm assuming the following workflow would work:

1. deepcopy the module.
2. Call `pyro.sample` for each prior and apply the inverse transform to get the appropriate value for the underlying raw parameter, as we do now.
3. Use `reset_params_pyro` above to replace the parameters with the samples on the copied module.

If so, I will probably implement things this way in the short term. It keeps most or all of the interface we already have for fully Bayesian inference but, you know, actually works 🙃 .

Regarding better support for `nn.Module` within Pyro's HMC: this would definitely be terrific to have. I imagine I'm not the only one that wants to take the parameters in a complex nested `nn.Module` and do HMC over them.

**Comment by neerajprad** [comment]
Your workflow sounds reasonable and hopefully should unblock you for the time being, but one thing to check would be if this works with the JIT. Let us know if you hit any other blockers as well. I can take a look at better supporting NN modules within HMC next week which should also be straightforward.

**Comment by fritzo** [comment]
@jacobrgardner another option might be to use [PyroModule](https://docs.pyro.ai/en/stable/nn.html#pyro.nn.module.PyroModule), which is the replacement for `pyro.random_module()`:
```py
# Define a torch.nn.Module.
model = ...

# Now make it Bayesian.
from pyro.nn.module import PyroSample, to_pyro_module_
to_pyro_module_(model)
model.covar_module.base_kernel.raw_lengthscale = PyroSample(Normal(0, 1).expand([1, 1]))
model.covar_module.raw_outputscale =  PyroSample(Normal(0, 1))
model.likelihood.noise_covar.raw_noise = PyroSample(Normal(0, 1).expand([1]))
model.mean_module.constant = PyroSample(Normal(0, 1))
```

**Comment by jacobrgardner** [comment]
@fritzo I initially tried that but couldn't get it working. For example:

```python
import math
import torch
import gpytorch
from gpytorch.kernels import RBFKernel, ScaleKernel
from gpytorch.kernels.rbf_kernel import postprocess_rbf
from gpytorch.means import ConstantMean
from torch.nn.functional import softplus

from gpytorch.functions import RBFCovariance

import pyro
from pyro.distributions import Normal
from pyro.infer.mcmc import NUTS, HMC, MCMC


class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

x = torch.linspace(0, 1, 10).double()
y = torch.sin(x * (2 * math.pi)).double() + torch.randn(x.size()).double() * 0.1

likelihood = gpytorch.likelihoods.GaussianLikelihood()
likelihood.double()

model = ExactGPModel(x, y, likelihood)
model.double()

from pyro.nn.module import PyroSample, to_pyro_module_
to_pyro_module_(model)

def pyro_model6(x, y):
    model.covar_module.base_kernel.raw_lengthscale = PyroSample(Normal(0, 1).expand([1, 1]))
    model.covar_module.raw_outputscale =  PyroSample(Normal(0, 1))
    model.likelihood.noise_covar.raw_noise = PyroSample(Normal(0, 1).expand([1]))
    model.mean_module.constant = PyroSample(Normal(0, 1))
    
    output = model.likelihood(model(x))
    pyro.sample("obs", output, obs=y)
```

Seems to not work as expected, possibly because of something we do in `gpytorch.Module`. If I run:
```python
tr = trace(pyro_model6).get_trace(x, y)
display(tr.nodes)
```
The only nodes that come back are `_INPUT`, `obs` and `_RETURN`. I think at some point I may have gotten slightly further than this, but maybe the potential gradients were still also wrong?

**Comment by fritzo** [comment]
@jacobrgardner thanks for explaining.  I wouldn't be surprised if there is some conflict between `PyroModule` and `gpytorch.Module`, especially as they have similar goals.  If @neerajprad solution works then great :+1:.  If you ever make further attempts at using `PyroModule`, don't hesitate to ping me on slack or on PRs.
